Add auto-generated connected components in documentation (#5791)
grafana/agent: docs/sources/flow/reference/compatibility/_index.md

Author: It does get a bit messy due to `string` being an odd one: we are only looking at components that export string and not listing those that take string as an argument, as that would be almost all components.

I would prefer to postpone improvements like this for future PRs. Doing this by hand may be a good option too - we can maintain a list of components that fetch an arbitrary string from somewhere.

> ---

Author: I intend to generate most of this file, especially the lists of components.

---

# Compatible components

This section provides an overview of _some_ of the possible connections between 
compatible components in Grafana Agent Flow. 

For each common telemetry data type, we provide a list of compatible components
that can export or consume it.

> Note that connecting some components may require further configuration to make
> the connection work correctly. Please refer to the linked documentation for more
> details.

## Targets

Targets are a `list(map(string))` - a [list]({{< relref "../../config-language/expressions/types_and_values/#naming-convention" >}}) of [maps]({{< relref "../../config-language/expressions/types_and_values/#naming-convention" >}}) with [string]({{< relref "../../config-language/expressions/types_and_values/#strings" >}}) values. As such,
they can contain different key-value pairs and can be used with a wide range of
components. Some components export Targets with key-value pairs specified in
the reference documentation, while other components accept Targets as arguments.
Some components require Targets to contain specific key-value pairs in order
to work correctly. It is recommended to always check component reference for

> details when working with Targets.

Author: Hmmm.... where else is it mentioned? I found the lack of clarity about what the targets are supposed to contain one of the big annoyances when getting started. Especially without prometheus background.

> details when working with Targets.

Author: I've simplified it a bit to capture the essence better.

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Targets Exporters
Below are components that _export_ Targets grouped by namespace. Click
on the namespace to expand and see more detail.

<!-- START GENERATED SECTION: EXPORTERS OF Targets -->

> {{< collapse title="discovery" >}}

Author: Not something I can do here, but we can probably request the UI team to help with this in a separate PR? cc @jdbaldry perhaps? Here's an illustration of the problem:
![image](https://github.com/grafana/agent/assets/17101802/d80c2428-0c64-47e3-bf76-4f7d9c1da082)

> {{< collapse title="discovery" >}}

Reviewer: Yeah it's currently locked in at a single size. It def could use a revisit to improve the way it is styled.

- [loki.source.kubernetes]({{< relref "../components/loki.source.kubernetes.md" >}})
{{< /collapse >}}

{{< collapse title="otelcol" >}}
- [otelcol.processor.discovery]({{< relref "../components/otelcol.processor.discovery.md" >}})
{{< /collapse >}}

{{< collapse title="prometheus" >}}
- [prometheus.scrape]({{< relref "../components/prometheus.scrape.md" >}})
{{< /collapse >}}

{{< collapse title="pyroscope" >}}
- [pyroscope.ebpf]({{< relref "../components/pyroscope.ebpf.md" >}})
- [pyroscope.scrape]({{< relref "../components/pyroscope.scrape.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: CONSUMERS OF Targets -->

## Prometheus `MetricsReceiver`

The Prometheus metrics are sent between components using `MetricsReceiver`s. 
`MetricsReceiver`s are [capsules]({{< relref "../../config-language/expressions/types_and_values/#capsules" >}})
that are exported by components that can receive Prometheus metrics. Components that
can consume Prometheus metrics can be passed the `MetricsReceiver` as an argument. Use the
following components to build your Prometheus metrics pipeline:

<!-- NOTE: this title is used as an anchor in links. Do not change. -->

> ### Prometheus `MetricsReceiver` Exporters

Author: This would break the anchors that we link to directly. I'm aware of the TOC, but I thought that it's more important to have good anchors 🤔 We could link to the parent heading though as an alternative...

- /docs/grafana-cloud/monitor-infrastructure/integrations/agent/flow/reference/compatible-components/
- /docs/grafana-cloud/send-data/agent/flow/reference/compatible-components/
canonical: https://grafana.com/docs/agent/latest/flow/reference/compatible-components/
description: Learn about which components are compatible with each other in Grafana Agent Flow
title: Compatible components
weight: 400
---

# Compatible components

This section provides an overview of _some_ of the possible connections between 
compatible components in Grafana Agent Flow. 

For each common data type, we provide a list of compatible components
that can export or consume it.

{{% admonition type="note" %}}

> The type of export may not be the only requirement for chaining components together.
> The value of an attribute may matter as well as its type.
> Please refer to each component's documentation for more details on what values are acceptable.
>
> For example:
> * A Prometheus component may always expect an `"__address__"` label inside a list of targets.
> * A `string` argument may only accept certain values like "traceID" or "spanID".

{{% /admonition %}}

## Targets

> Targets are a `list(map(string))` - a [list]({{< relref "../../config-language/expressions/types_and_values/#naming-convention" >}}) of [maps]({{< relref "../../config-language/expressions/types_and_values/#naming-convention" >}}) with [string]({{< relref "../../config-language/expressions/types_and_values/#strings" >}}) values. As such,
> they can contain different key-value pairs and can be used with a wide range of
> components. Some components require Targets to contain specific key-value pairs in order
> to work correctly, it is recommended to always check component reference for
> details when working with Targets.

Reviewer: ```suggestion
Targets are a `list(map(string))` - a [list]({{< relref "../../config-language/expressions/types_and_values/#naming-convention" >}}) of [maps]({{< relref "../../config-language/expressions/types_and_values/#naming-convention" >}}) with [string]({{< relref "../../config-language/expressions/types_and_values/#strings" >}}) values.
They can contain different key-value pairs, and you can use them with a wide range of
components. Some components require Targets to contain specific key-value pairs
to work correctly. It is recommended to always check component references for
details when working with Targets.
```

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Targets Exporters

> Below are components that _export_ Targets grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace,  _export_ Targets.
```

I think we can remove the `Click to expand` part.

- [prometheus.exporter.oracledb]({{< relref "../components/prometheus.exporter.oracledb.md" >}})
- [prometheus.exporter.postgres]({{< relref "../components/prometheus.exporter.postgres.md" >}})
- [prometheus.exporter.process]({{< relref "../components/prometheus.exporter.process.md" >}})
- [prometheus.exporter.redis]({{< relref "../components/prometheus.exporter.redis.md" >}})
- [prometheus.exporter.snmp]({{< relref "../components/prometheus.exporter.snmp.md" >}})
- [prometheus.exporter.snowflake]({{< relref "../components/prometheus.exporter.snowflake.md" >}})
- [prometheus.exporter.squid]({{< relref "../components/prometheus.exporter.squid.md" >}})
- [prometheus.exporter.statsd]({{< relref "../components/prometheus.exporter.statsd.md" >}})
- [prometheus.exporter.unix]({{< relref "../components/prometheus.exporter.unix.md" >}})
- [prometheus.exporter.vsphere]({{< relref "../components/prometheus.exporter.vsphere.md" >}})
- [prometheus.exporter.windows]({{< relref "../components/prometheus.exporter.windows.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: EXPORTERS OF Targets -->

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Targets Consumers

> Below are components that _consume_ Targets grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _consume_ Targets.
```

{{< /collapse >}}

{{< collapse title="otelcol" >}}
- [otelcol.processor.discovery]({{< relref "../components/otelcol.processor.discovery.md" >}})
{{< /collapse >}}

{{< collapse title="prometheus" >}}
- [prometheus.scrape]({{< relref "../components/prometheus.scrape.md" >}})
{{< /collapse >}}

{{< collapse title="pyroscope" >}}
- [pyroscope.ebpf]({{< relref "../components/pyroscope.ebpf.md" >}})
- [pyroscope.scrape]({{< relref "../components/pyroscope.scrape.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: CONSUMERS OF Targets -->

## Prometheus `MetricsReceiver`

The Prometheus metrics are sent between components using `MetricsReceiver`s. 
`MetricsReceiver`s are [capsules]({{< relref "../../config-language/expressions/types_and_values/#capsules" >}})
that are exported by components that can receive Prometheus metrics. Components that
can consume Prometheus metrics can be passed the `MetricsReceiver` as an argument. Use the
following components to build your Prometheus metrics pipeline:

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Prometheus `MetricsReceiver` Exporters

> Below are components that _export_ Prometheus `MetricsReceiver`, grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _export_ Prometheus `MetricsReceiver`.
```

<!-- START GENERATED SECTION: EXPORTERS OF Prometheus `MetricsReceiver` -->

{{< collapse title="otelcol" >}}
- [otelcol.receiver.prometheus]({{< relref "../components/otelcol.receiver.prometheus.md" >}})
{{< /collapse >}}

{{< collapse title="prometheus" >}}
- [prometheus.relabel]({{< relref "../components/prometheus.relabel.md" >}})
- [prometheus.remote_write]({{< relref "../components/prometheus.remote_write.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: EXPORTERS OF Prometheus `MetricsReceiver` -->

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Prometheus `MetricsReceiver` Consumers

> Below are components that _consume_ Prometheus `MetricsReceiver`, grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _consume_ Prometheus `MetricsReceiver`.
```

{{< /collapse >}}

{{< collapse title="prometheus" >}}
- [prometheus.operator.podmonitors]({{< relref "../components/prometheus.operator.podmonitors.md" >}})
- [prometheus.operator.probes]({{< relref "../components/prometheus.operator.probes.md" >}})
- [prometheus.operator.servicemonitors]({{< relref "../components/prometheus.operator.servicemonitors.md" >}})
- [prometheus.receive_http]({{< relref "../components/prometheus.receive_http.md" >}})
- [prometheus.relabel]({{< relref "../components/prometheus.relabel.md" >}})
- [prometheus.scrape]({{< relref "../components/prometheus.scrape.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: CONSUMERS OF Prometheus `MetricsReceiver` -->

## Loki `LogsReceiver`

`LogsReceiver` is a [capsule]({{< relref "../../config-language/expressions/types_and_values/#capsules" >}})
that is exported by components that can receive Loki logs. Components that
consume `LogsReceiver` as an argument typically send logs to it. Use the
following components to build your Loki logs pipeline:

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Loki `LogsReceiver` Exporters

> Below are components that _export_ Loki `LogsReceiver` grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _export_ Loki `LogsReceiver`.
```

<!-- START GENERATED SECTION: EXPORTERS OF Loki `LogsReceiver` -->

{{< collapse title="loki" >}}
- [loki.echo]({{< relref "../components/loki.echo.md" >}})
- [loki.process]({{< relref "../components/loki.process.md" >}})
- [loki.relabel]({{< relref "../components/loki.relabel.md" >}})
- [loki.write]({{< relref "../components/loki.write.md" >}})
{{< /collapse >}}

{{< collapse title="otelcol" >}}
- [otelcol.receiver.loki]({{< relref "../components/otelcol.receiver.loki.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: EXPORTERS OF Loki `LogsReceiver` -->

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Loki `LogsReceiver` Consumers

> Below are components that _consume_ Loki `LogsReceiver` grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _consume_ Loki `LogsReceiver`.
```

- [loki.source.kafka]({{< relref "../components/loki.source.kafka.md" >}})
- [loki.source.kubernetes]({{< relref "../components/loki.source.kubernetes.md" >}})
- [loki.source.kubernetes_events]({{< relref "../components/loki.source.kubernetes_events.md" >}})
- [loki.source.podlogs]({{< relref "../components/loki.source.podlogs.md" >}})
- [loki.source.syslog]({{< relref "../components/loki.source.syslog.md" >}})
- [loki.source.windowsevent]({{< relref "../components/loki.source.windowsevent.md" >}})
{{< /collapse >}}

{{< collapse title="otelcol" >}}
- [otelcol.exporter.loki]({{< relref "../components/otelcol.exporter.loki.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: CONSUMERS OF Loki `LogsReceiver` -->

## OpenTelemetry `otelcol.Consumer`

The OpenTelemetry data is sent between components using `otelcol.Consumer`s.
`otelcol.Consumer`s are [capsules]({{< relref "../../config-language/expressions/types_and_values/#capsules" >}})
that are exported by components that can receive OpenTelemetry data. Components that
can consume OpenTelemetry data can be passed the `otelcol.Consumer` as an argument. Note that some components

> that use `otelcol.Consumer` only support a subset of telemetry signals, e.g. only traces. Check the component

Reviewer: ```suggestion
that use `otelcol.Consumer` only support a subset of telemetry signals, for example, only traces. Check the component
```

reference pages for more details on what is supported. Use the following components to build your OpenTelemetry pipeline:

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### OpenTelemetry `otelcol.Consumer` Exporters

> Below are components that _export_ OpenTelemetry `otelcol.Consumer`, grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _export_ OpenTelemetry `otelcol.Consumer`.
```

- [otelcol.processor.attributes]({{< relref "../components/otelcol.processor.attributes.md" >}})
- [otelcol.processor.batch]({{< relref "../components/otelcol.processor.batch.md" >}})
- [otelcol.processor.discovery]({{< relref "../components/otelcol.processor.discovery.md" >}})
- [otelcol.processor.filter]({{< relref "../components/otelcol.processor.filter.md" >}})
- [otelcol.processor.k8sattributes]({{< relref "../components/otelcol.processor.k8sattributes.md" >}})
- [otelcol.processor.memory_limiter]({{< relref "../components/otelcol.processor.memory_limiter.md" >}})
- [otelcol.processor.probabilistic_sampler]({{< relref "../components/otelcol.processor.probabilistic_sampler.md" >}})
- [otelcol.processor.span]({{< relref "../components/otelcol.processor.span.md" >}})
- [otelcol.processor.tail_sampling]({{< relref "../components/otelcol.processor.tail_sampling.md" >}})
- [otelcol.processor.transform]({{< relref "../components/otelcol.processor.transform.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: EXPORTERS OF OpenTelemetry `otelcol.Consumer` -->

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### OpenTelemetry `otelcol.Consumer` Consumers

> Below are components that _consume_ OpenTelemetry `otelcol.Consumer`, grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _consume_ OpenTelemetry `otelcol.Consumer`.
```

- [otelcol.receiver.opencensus]({{< relref "../components/otelcol.receiver.opencensus.md" >}})
- [otelcol.receiver.otlp]({{< relref "../components/otelcol.receiver.otlp.md" >}})
- [otelcol.receiver.prometheus]({{< relref "../components/otelcol.receiver.prometheus.md" >}})
- [otelcol.receiver.vcenter]({{< relref "../components/otelcol.receiver.vcenter.md" >}})
- [otelcol.receiver.zipkin]({{< relref "../components/otelcol.receiver.zipkin.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: CONSUMERS OF OpenTelemetry `otelcol.Consumer` -->

## Pyroscope `ProfilesReceiver`

The Pyroscope profiles are sent between components using `ProfilesReceiver`s.
`ProfilesReceiver`s are [capsules]({{< relref "../../config-language/expressions/types_and_values/#capsules" >}})
that are exported by components that can receive Pyroscope profiles. Components that
can consume Pyroscope profiles can be passed the `ProfilesReceiver` as an argument. Use the
following components to build your Pyroscope profiles pipeline:

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Pyroscope `ProfilesReceiver` Exporters

> Below are components that _export_ Pyroscope `ProfilesReceiver`, grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _export_ Pyroscope `ProfilesReceiver`.
```

<!-- START GENERATED SECTION: EXPORTERS OF Pyroscope `ProfilesReceiver` -->

{{< collapse title="pyroscope" >}}
- [pyroscope.write]({{< relref "../components/pyroscope.write.md" >}})
{{< /collapse >}}

<!-- END GENERATED SECTION: EXPORTERS OF Pyroscope `ProfilesReceiver` -->

<!-- NOTE: this title is used as an anchor in links. Do not change. -->
### Pyroscope `ProfilesReceiver` Consumers

> Below are components that _consume_ Pyroscope `ProfilesReceiver`, grouped by namespace. Click
> on the namespace to expand and see more detail.

Reviewer: ```suggestion
The following components, grouped by namespace, _consume_ Pyroscope `ProfilesReceiver`.
```
